tests/multigrid/transfer_matrix_free_device_01.cc

```diff
@@ -0,0 +1,190 @@
+// series of meshes with uniform meshes for FE_Q
+
+#include <deal.II/distributed/tria.h>
+
+#include <deal.II/fe/fe_q.h>
+
+#include <deal.II/grid/grid_generator.h>
+
+#include <deal.II/lac/la_parallel_vector.h>
+
+#include <deal.II/multigrid/mg_base.h>
+#include <deal.II/multigrid/mg_base.templates.h>
+#include <deal.II/multigrid/mg_transfer.h>
+#include <deal.II/multigrid/mg_transfer_matrix_free.h>
+
+#include "../tests.h"
+
+using namespace dealii;
+
+template <typename Number2>
+void
+copy_to_host(
+  LinearAlgebra::distributed::Vector<Number2, dealii::MemorySpace::Host>  &dst,
+  const LinearAlgebra::distributed::Vector<Number2, MemorySpace::Default> &src)
+{
+  LinearAlgebra::ReadWriteVector<Number2> rw_vector(
+    src.get_partitioner()->locally_owned_range());
+  rw_vector.import_elements(src, VectorOperation::insert);
+
+  dst.reinit(src.get_partitioner());
+  dst.import_elements(rw_vector, VectorOperation::insert);
+}
+
+template <typename Number2>
+void
+copy_from_host(
+  LinearAlgebra::distributed::Vector<Number2, MemorySpace::Default> &dst,
+  const LinearAlgebra::distributed::Vector<Number2, dealii::MemorySpace::Host>
+    &src)
+{
+  LinearAlgebra::ReadWriteVector<Number2> rw_vector(
+    src.get_partitioner()->locally_owned_range());
+  rw_vector.import_elements(src, VectorOperation::insert);
+
+  if (dst.size() == 0)
+    dst.reinit(src.get_partitioner());
+  dst.import_elements(rw_vector, VectorOperation::insert);
+}
+
+template <int dim, typename Number>
+void
+check(const unsigned int fe_degree)
+{
+  FE_Q<dim> fe(fe_degree);
+  deallog << "FE: " << fe.get_name() << std::endl;
+
+  // run a few different sizes...
+  unsigned int sizes[] = {1, 2, 3, 4, 5, 6, 8};
+  for (unsigned int cycle = 0; cycle < sizeof(sizes) / sizeof(unsigned int);
+       ++cycle)
+    {
+      unsigned int n_refinements = 0;
+      unsigned int n_subdiv      = sizes[cycle];
+      if (n_subdiv > 1)
+        while (n_subdiv % 2 == 0)
+          {
+            n_refinements += 1;
+            n_subdiv /= 2;
+          }
+      n_refinements += 3 - dim;
+      if (fe_degree < 3)
+        n_refinements += 1;
+      parallel::distributed::Triangulation<dim> tr(
+        MPI_COMM_WORLD,
+        Triangulation<dim>::limit_level_difference_at_vertices,
+        parallel::distributed::Triangulation<
+          dim>::construct_multigrid_hierarchy);
+      GridGenerator::subdivided_hyper_cube(tr, n_subdiv);
+      tr.refine_global(n_refinements);
+      deallog << "no. cells: " << tr.n_global_active_cells() << std::endl;
+
+      DoFHandler<dim> mgdof(tr);
+      mgdof.distribute_dofs(fe);
+      mgdof.distribute_mg_dofs();
+
+      MGConstrainedDoFs mg_constrained_dofs;
+      mg_constrained_dofs.initialize(mgdof);
+      mg_constrained_dofs.make_zero_boundary_constraints(mgdof, {0});
+
+      // build host reference
+      MGTransferMatrixFree<dim, Number, dealii::MemorySpace::Host>
+        transfer_host(mg_constrained_dofs);
```

Comment:
It looks like you are using local smoothing multi grid here. Not wrong, but portable MatrixFree is currently only doing global coarsening.